File: apps/reading/src/api/htmlNativeViewSessionAuthorityCompose.ts
/**
 * HTML-native view session authority pack (pure).
 *
 * Operator vision: every information asset viewed as HTML — reading and
 * research share the same authority doctrine; open a session only when HTML
 * projection is ready; PDF never primary.
 *
 * pdf_view_authorized always false.
 * pdf_primary always false.
 * store_mutated always false.
 */

import {
  composeHtmlAssetViewSession,
  type HtmlAssetViewSessionCompose,
} from "./htmlAssetViewSessionCompose";
import {
  composeReadingResearchHtmlParity,
  type ModeAssetViewInput,
  type ReadingResearchHtmlParityCompose,
} from "./readingResearchHtmlParityCompose";
import {
  evaluateHtmlNativeViewAuthority,
  type HtmlNativeViewAuthorityDecision,
} from "./htmlNativeViewAuthority";

export interface HtmlNativeViewSessionAuthorityInput {
  session_id: string;
  /** Primary asset for the open session. */
  asset_id: string;
  html_projection_sha: string | null;
  view_requested: boolean;
  twin_bound: boolean;
  twin_substrate_ready?: boolean;
  claimed_format?: string | null;
  /**
   * Reading/research parity surface inputs. When omitted, both modes use
   * the primary asset with source_format derived from claimed_format.
   */
  reading?: ModeAssetViewInput | null;
  research?: ModeAssetViewInput | null;
  operator_ack: boolean;
}

export interface HtmlNativeViewSessionAuthorityCompose {
  session_id: string;
  asset_id: string;
  session: HtmlAssetViewSessionCompose;
  authority: HtmlNativeViewAuthorityDecision;
  parity: ReadingResearchHtmlParityCompose;
  /**
   * True when session.session_ready, authority.human_viewable_html,
   * parity.parity_ready (or both unavailable honesty), and operator_ack.
   */
  pack_ready: boolean;
  pdf_view_authorized: false;
  pdf_primary: false;
  store_mutated: false;
  notes: string[];
  pack_authority: "html_native_view_session_authority_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function deriveFormat(
  claimed: string | null | undefined,
): ModeAssetViewInput["source_format"] {
  if (claimed == null || claimed === undefined || !String(claimed).trim()) {
    return "unknown";
  }
  const f = String(claimed).trim().toLowerCase();
  if (f === "html" || f === "pdf" || f === "epub" || f === "markdown") {
    return f;
  }
  return "unknown";
}

/**
 * Compose HTML asset session + authority + reading/research parity.
 * Never authorizes PDF primary; never mutates store.
 */
export function composeHtmlNativeViewSessionAuthority(
  input: HtmlNativeViewSessionAuthorityInput,
): HtmlNativeViewSessionAuthorityCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const asset_id = requireNonEmpty(input.asset_id, "asset_id");

  const notes: string[] = [
    "pdf_view_authorized=false — HTML-native doctrine",
    "pdf_primary=false",
    "store_mutated=false",
  ];

  const session = composeHtmlAssetViewSession({
    session_id,
    asset_id,
    html_projection_sha: input.html_projection_sha,
    view_requested: input.view_requested,
    twin_bound: input.twin_bound,
    twin_substrate_ready: input.twin_substrate_ready,
    claimed_format: input.claimed_format,
  });
  notes.push(...session.notes.map((n) => `[session] ${n}`));

  const source_format = deriveFormat(input.claimed_format);
  const authority = evaluateHtmlNativeViewAuthority({
    asset_id,
    asset_kind: "research",
    source_format,
    html_projection_sha: input.html_projection_sha,
    prefer_html: true,
    allow_pdf_secondary: false,
  });
  notes.push(...authority.notes.map((n) => `[authority] ${n}`));

  const reading: ModeAssetViewInput =
    input.reading != null
      ? input.reading
      : {
          asset_id,
          asset_kind: "book",
          source_format,
          html_projection_sha: input.html_projection_sha,
          prefer_html: true,
          allow_pdf_secondary: false,
        };
  const research: ModeAssetViewInput =
    input.research != null
      ? input.research
      : {
          asset_id,
          asset_kind: "research",
          source_format,
          html_projection_sha: input.html_projection_sha,
          prefer_html: true,
          allow_pdf_secondary: false,
        };

  const parity = composeReadingResearchHtmlParity({ reading, research });
  notes.push(...parity.notes.map((n) => `[parity] ${n}`));

  const pack_ready =
    session.session_ready === true &&
    authority.human_viewable_html === true &&
    parity.pdf_primary === false &&
    session.pdf_view_authorized === false &&
    input.operator_ack === true;

  if (pack_ready) {
    notes.push(
      "pack_ready=true — HTML view session + authority + parity ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — session, authority HTML, or operator_ack gate open",
    );
  }

  if (
    session.pdf_view_authorized !== false ||
    session.store_mutated !== false ||
    parity.pdf_primary !== false
  ) {
    throw new Error("invariant: PDF must never be authorized as primary view");
  }

  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("store_mutated=false");

  return {
    session_id,
    asset_id,
    session,
    authority,
    parity,
    pack_ready,
    pdf_view_authorized: false,
    pdf_primary: false,
    store_mutated: false,
    notes,
    pack_authority: "html_native_view_session_authority_compose_advisory",
  };
}

export function formatHtmlNativeViewSessionAuthoritySummary(
  c: HtmlNativeViewSessionAuthorityCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · session_ready=${c.session.session_ready} · ` +
    `human_viewable_html=${c.authority.human_viewable_html} · ` +
    `both_html_ready=${c.parity.both_html_ready} · ` +
    `pdf_view_authorized=false · pdf_primary=false · store_mutated=false`
  );
}

File: substrate/html_native_view_session_authority_compose.py
"""HTML-native view session authority pack (pure).

pdf_view_authorized, pdf_primary, store_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.html_asset_view_session_compose import (
    HtmlAssetViewSessionCompose,
    HtmlAssetViewSessionComposeError,
    compose_html_asset_view_session,
)
from substrate.html_native_view_authority import (
    HtmlNativeViewAuthorityDecision,
    HtmlNativeViewAuthorityError,
    evaluate_html_native_view_authority,
)
from substrate.reading_research_html_parity_compose import (
    ReadingResearchHtmlParityCompose,
    ReadingResearchHtmlParityComposeError,
    compose_reading_research_html_parity,
)


class HtmlNativeViewSessionAuthorityComposeError(ValueError):
    """Fail-closed validation for HTML view session authority pack."""


@dataclass(frozen=True)
class HtmlNativeViewSessionAuthorityCompose:
    session_id: str
    asset_id: str
    session: HtmlAssetViewSessionCompose
    authority: HtmlNativeViewAuthorityDecision
    parity: ReadingResearchHtmlParityCompose
    pack_ready: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    store_mutated: bool
    notes: tuple[str, ...]
    pack_authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "asset_id": self.asset_id,
            "session": self.session.to_dict(),
            "authority": self.authority.to_dict(),
            "parity": self.parity.to_dict(),
            "pack_ready": self.pack_ready,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "pack_authority": (
                "html_native_view_session_authority_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise HtmlNativeViewSessionAuthorityComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _derive_format(claimed: object | None) -> str:
    if claimed is None or (isinstance(claimed, str) and not claimed.strip()):
        return "unknown"
    f = str(claimed).strip().lower()
    if f in ("html", "pdf", "epub", "markdown"):
        return f
    return "unknown"


def compose_html_native_view_session_authority(
    *,
    session_id: object,
    asset_id: object,
    html_projection_sha: object,
    view_requested: object,
    twin_bound: object,
    operator_ack: object,
    twin_substrate_ready: object | None = None,
    claimed_format: object | None = None,
    reading: object | None = None,
    research: object | None = None,
) -> HtmlNativeViewSessionAuthorityCompose:
    """Session + authority + parity. Never PDF primary; never mutates store."""
    if not isinstance(operator_ack, bool):
        raise HtmlNativeViewSessionAuthorityComposeError(
            "operator_ack must be an explicit boolean"
        )
    sid = _require_nonempty(session_id, field="session_id")
    aid = _require_nonempty(asset_id, field="asset_id")

    notes: list[str] = [
        "pdf_view_authorized=false — HTML-native doctrine",
        "pdf_primary=false",
        "store_mutated=false",
    ]

    try:
        session = compose_html_asset_view_session(
            session_id=sid,
            asset_id=aid,
            html_projection_sha=html_projection_sha,
            view_requested=view_requested,
            twin_bound=twin_bound,
            twin_substrate_ready=twin_substrate_ready,
            claimed_format=claimed_format,
        )
    except HtmlAssetViewSessionComposeError as e:
        raise HtmlNativeViewSessionAuthorityComposeError(str(e)) from e
    notes.extend(f"[session] {n}" for n in session.notes)

    source_format = _derive_format(claimed_format)
    try:
        authority = evaluate_html_native_view_authority(
            asset_id=aid,
            asset_kind="research",
            source_format=source_format,
            html_projection_sha=html_projection_sha,
            prefer_html=True,
            allow_pdf_secondary=False,
        )
    except HtmlNativeViewAuthorityError as e:
        raise HtmlNativeViewSessionAuthorityComposeError(str(e)) from e
    notes.extend(f"[authority] {n}" for n in authority.notes)

    reading_in: dict[str, Any]
    research_in: dict[str, Any]
    if reading is not None:
        if not isinstance(reading, dict):
            raise HtmlNativeViewSessionAuthorityComposeError(
                "reading must be an object when set"
            )
        reading_in = dict(reading)
    else:
        reading_in = {
            "asset_id": aid,
            "asset_kind": "book",
            "source_format": source_format,
            "html_projection_sha": html_projection_sha,
            "prefer_html": True,
            "allow_pdf_secondary": False,
        }
    if research is not None:
        if not isinstance(research, dict):
            raise HtmlNativeViewSessionAuthorityComposeError(
                "research must be an object when set"
            )
        research_in = dict(research)
    else:
        research_in = {
            "asset_id": aid,
            "asset_kind": "research",
            "source_format": source_format,
            "html_projection_sha": html_projection_sha,
            "prefer_html": True,
            "allow_pdf_secondary": False,
        }

    try:
        parity = compose_reading_research_html_parity(
            reading=reading_in,
            research=research_in,
        )
    except ReadingResearchHtmlParityComposeError as e:
        raise HtmlNativeViewSessionAuthorityComposeError(str(e)) from e
    notes.extend(f"[parity] {n}" for n in parity.notes)

    pack_ready = (
        session.session_ready is True
        and authority.human_viewable_html is True
        and parity.pdf_primary is False
        and session.pdf_view_authorized is False
        and operator_ack is True
    )
    if pack_ready:
        notes.append(
            "pack_ready=true — HTML view session + authority + parity ready; "
            "still pure"
        )
    else:
        notes.append(
            "pack_ready=false — session, authority HTML, or operator_ack gate open"
        )

    if (
        session.pdf_view_authorized is not False
        or session.store_mutated is not False
        or parity.pdf_primary is not False
    ):
        raise HtmlNativeViewSessionAuthorityComposeError(
            "invariant: PDF must never be authorized as primary view"
        )

    notes.extend(
        (
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "store_mutated=false",
        )
    )

    return HtmlNativeViewSessionAuthorityCompose(
        session_id=sid,
        asset_id=aid,
        session=session,
        authority=authority,
        parity=parity,
        pack_ready=pack_ready,
        pdf_view_authorized=False,
        pdf_primary=False,
        store_mutated=False,
        notes=tuple(notes),
        pack_authority="html_native_view_session_authority_compose_advisory",
    )


def format_html_native_view_session_authority_summary(
    c: HtmlNativeViewSessionAuthorityCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · session_ready={c.session.session_ready} · "
        f"human_viewable_html={c.authority.human_viewable_html} · "
        f"both_html_ready={c.parity.both_html_ready} · "
        f"pdf_view_authorized=false · pdf_primary=false · store_mutated=false"
    )


__all__ = [
    "HtmlNativeViewSessionAuthorityCompose",
    "HtmlNativeViewSessionAuthorityComposeError",
    "compose_html_native_view_session_authority",
    "format_html_native_view_session_authority_summary",
]
